Scene v2

PhaestusFox opened the pull request:
have been procrastinating on a different project this weekend, so have started working on a fully reflective-based scene serialiser.

Currently, it can only serialise into a file, and I know deserialising is going to be so, so much harder.
figured I would open a pull so you guys can have a look and give any input, probably won't work on it much till next weekend.

gone for a bit of a hybrid between Toml and Ron.
each section is split by [header]
followed by "entries" that are sort of one per line that work in a key: value layout
In the [entities] section, the keys are the type name of the component, and the value is the serialised data.

The hope is that it shouldn't be too hard to get the type name, then, with reflection, recursively rebuild the data.
The semi-colon at the end is about making it easier to break the file up into chunks, since you can just scan till you find a semicolon, counting quotes to make sure we are not in a String.

Something else I'm considering is having a "reduced" format and a "compact" format.
Reduced: will check to see if a component/value has its default value, and if so, will not save that to the file; this will hopefully cut down on bloat/make it easier to hand-modify scene files since missing fields will be assumed default.
Compact: will use the "short" names of a type if they don't conflict with any other types, as well as all improvements added on Reduced. The big drawback to this is it can make scene files a lot more fragile since if any dependency ever registers a component with the same name, the short name becomes invalid and the scene will have no way of unambiubating the types.

There is also a way to flag components as not to be serialised, for things like GlobalTransform that should be constructed by all the Transforms at runtime, never stored directly.
You can also flag whole entities to be skipped.
I have also implemented a reflect_trait that lets you override how something is stored, this I only used for Name at the moment because I don't want to serialise the has representation, and there is no obviuse way to me to skip it without modifying Bevy to somehow mark that field as unimportant with say #[serde(ignore)] so hard code it directily is what I did.

I have kept all entities being stored as UUIDs, but we should probably store the UUID as a component when we load a scene, to maybe use for debugging or hot reloading of scenes

PhaestusFox committed:
make transform gizmo only use primary InputButton

Technically any two button clicks on the gizmo gets it in an invalid state since two sets of Drag events are sent and are in different axis Right mouse is just not as obvious because the camera is moving and breaks the raycast
Ps. Is there any reason we are using UserInput instead of bevy's input system?
feels like redundant work to maintain and prone to missmatch logic It also makes the editor less versatile since things like picking have a abstracted concept of "Primary" and "Secondary" button so it works independently of user control surface. eg touch, pen, mouse

PhaestusFox committed:
cleaned up rotation gizmo code

PhaestusFox committed:
only allow rotation gizmo to work with primary click

PhaestusFox committed:
cleaned up rotation gizmo code

PhaestusFox committed:
added customizable setting per gizmo

PhaestusFox committed:
stop print every fream

PhaestusFox committed:
make the editor able to load any supported scene versions

PhaestusFox committed:
added reflect serializer

PhaestusFox committed:
Merge branch 'dev' into scene-v2

BlakeDarrow commented:
Very interested in this! Would love to learn more. What are your main goals? I do immediately see a huge benefit of reducing the file size and complexity of a scene, and I do like the notion of making the file easier to hand-edit!

PhaestusFox commented:
The main goal is to make it pure reflection, this is so we don't have to worry about if the user implements serde or whatnot, since reflect is a fairly common derive when working in some areas of bevy, it's also something bevy seems pretty determined to implement on all its types, even some that doent make any sense like Frustom that litteraly just ignores its only field and is therefor compleatly useless outside of type name.

I also think Reflect is more robust than serde, since even if they don't directly implement certain behaviours or attributes, by using the type registry, we can dynamically decide how to handle things like missing fields.

Rust effectively only has 3 types: Int, float, and strings. With Bevy, there is a fourth Entity, but this is only because we want to represent that with a UUID, and so requires special logic.

With these three types that Reflect calls opaque types, basically, too small to break down further, we can construct any other reflected type. This should keep the implementation small and highly recursive.

I also would like to keep it reading like Rust code, like we are instantiating a struct.

on a feature creep side note.
I was also thinking it might be possible to include constructior functions, and named constants as another way to cut down on the size of a scene file.
something like:
Transform: IDENTITY
vs
Transform: {
		translation: {
			x: 0.99008465,
			y: 0.40590173,
			z: 0.9690189,
		},
		rotation: {
			x: 0.0,
			y: 0.0,
			z: 0.0,
			w: 1.0,
		},
		scale: {
			x: 1.0,
			y: 1.0,
			z: 1.0,
		},
	};
and 
Camera3d: default()
vs 
Camera3d: {
		depth_load_op: Clear(0.0),
		depth_texture_usages: 16,
		screen_space_specular_transmission_steps: 1,
		screen_space_specular_transmission_quality: Medium,
	};

PhaestusFox committed:
start adding reduced format

PhaestusFox committed:
Merge branch 'scene-v2' of https://github.com/PhaestusFox/bevy_granite into scene-v2

PhaestusFox committed:
added reduced format

the reduced format will use the type name only if it can.
if the type name is already taken it will make one step up the module path and use that, repeat till available name is found.
at the end of the scene file it adds a list of all name mappings.
{short} -> {full}

I could use the type_regustry to lookup if a name is_ambiguse, this just means two registered types share a name, then use the short name if not of the full name if, but I think for compatibility reasons a scene should keep note of the full name otherwise if you add a dependency that has a conflicting type there is no way of knowing what the original was. it also means that we don't have to fall back to full paths to resolve conflicts we can just step one module up

BlakeDarrow commented:
I made some changes to the **RequestLoadEvent** where it takes a **SaveSettings** enum and optional **transform**. This allows us the ability to tag our loaded world with how we want it to handle saving, and offset the loading position. 

If you are having issues with conflicts, lmk when your done and I can merge this into another branch from before and fix the conflicts

PhaestusFox committed:
I dont remember doing this

BlakeDarrow committed:
v0.2.1 release (#77)

Merge pull request #77 from BlakeDarrow/dev

BlakeDarrow committed:
v0.2.2 release (#81)

* feat(camera): Add dedicated editor camera (#78)

- Spawn a persistent editor-specific Camera3d and track its state with
ViewportCameraState
- Keep the UI/gizmo camera in sync with whichever camera is active (editor or
scene override)
- Allow opting into scene cameras via UI and restore runtime MainCameras when
leaving edit mode

* bump version v0.2.2 (#79)

---------

Co-authored-by: drif <mitchell.boulter@googlemail.com>

PhaestusFox committed:
Merge branch 'main' into scene-v2
